static/app/components/asyncComponentSearchInput.tsx

```diff
@@ -15,138 +16,200 @@ type RenderProps = {
-  };
-
-type State = {
-  busy: boolean;
-  query: string;
-};
+  className?: string;
+  /**
+   * Time in milliseconds to wait before firing off the request
+   */
+  debounceWait?: number;
+  onSearchSubmit?: (query: string, event: React.FormEvent) => void;
 
-/**
- * This is a search input that can be easily used in AsyncComponent/Views.
- *
- * It probably doesn't make too much sense outside of an AsyncComponent atm.
- */
-class AsyncComponentSearchInput extends React.Component<Props, State> {
-  static defaultProps: DefaultProps = {
-    placeholder: t('Search...'),
-    debounceWait: 200,
-  };
+  /**
+   * Updates URL with search query in the URL param: `query`
+   */
+  updateRoute?: boolean;
+}
 
-  state: State = {
+function AsyncComponentSearchInput({
+  placeholder = t('Search...'),
+  debounceWait = 200,
+  api,
+  className,
+  url,
+  updateRoute,
+  onSearchSubmit,
+  onSuccess,
+  onError,
+  router,
+  location,
+  children,
+}: AsyncComponentSearchInputProps): React.ReactElement {
+  const [state, setState] = React.useState<{busy: boolean; query: string}>({
     query: '',
     busy: false,
-  };
-
-  immediateQuery = async (searchQuery: string) => {
-    const {location, api} = this.props;
-    this.setState({busy: true});
-
-    try {
-      const [data, , resp] = await api.requestPromise(`${this.props.url}`, {
-        includeAllArgs: true,
-        method: 'GET',
-        query: {...location.query, query: searchQuery},
-      });
-      // only update data if the request's query matches the current query
-      if (this.state.query === searchQuery) {
-        this.props.onSuccess(data, resp);
-      }
-    } catch {
-      this.props.onError();
-    }
-
-    this.setState({busy: false});
-  };
-
-  query = debounce(this.immediateQuery, this.props.debounceWait);
-
-  handleChange = (query: string) => {
-    this.query(query);
-    this.setState({query});
-  };
-
-  handleInputChange = (evt: React.ChangeEvent<HTMLInputElement>) =>
-    this.handleChange(evt.target.value);
+  });
+
+  // We need to use a mutable ref to keep reference to our latest query, else
+  // useCallback scope will always reference the previous value and our condition will always fail
+  const latestQuery = React.useRef<string>(state.query);
+
+  const queryResolver = React.useCallback(
+    (searchQuery: string) => {
+      setState({query: searchQuery, busy: true});
+
+      api
+        .requestPromise(url, {
+          includeAllArgs: true,
+          method: 'GET',
+          query: {...location.query, query: searchQuery},
+        })
+        .then(([data, , resp]) => {
+          if (latestQuery.current === searchQuery) {
+            onSuccess(data, resp);
+          }
+        })
+        .catch(() => {
+          if (latestQuery.current === searchQuery) {
+            onError();
+          }
+        })
+        .finally(() => {
+          if (latestQuery.current === searchQuery) {
+            setState({query: searchQuery, busy: false});
```

Comment:
shouldn't we always set the busy to false independently of the condition?

Reply:
I will be set when we eventually hit promise.finally, I think meanwhile it makes sense to keep it busy to indicate that we are searching while the user types.